File: typescript/packages/management/src/specs.ts
import { readFileSync, readdirSync, existsSync, statSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

declare const __dirname: string | undefined;

export const MIN_SPEC_VERSION = 'v1';
export const MAX_SPEC_VERSION = 'v1';

export interface ManagementSpec {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  outputSchema: Record<string, unknown>;
  since: string;
}

export interface ManagementFixture {
  input: Record<string, unknown>;
  output: Record<string, unknown>;
}

export interface LoadedManagementCorpus {
  specDir: string;
  specs: Record<string, ManagementSpec>;
  fixtures: Record<string, ManagementFixture>;
}

export interface LoadOptions {
  specDir?: string;
}

const ENV_OVERRIDE = 'COPASS_MANAGEMENT_SPEC_DIR';

function resolveModuleDir(): string {
  // CJS bundles inject `__dirname` directly; prefer it when present so we
  // don't trip the `import.meta` constraint in `verbatimModuleSyntax`.
  if (typeof __dirname === 'string') return __dirname;
  // ESM path: read `import.meta.url` via an indirect lookup so TS doesn't
  // require `module: esnext` for the file (it's still emitted as ESM).
  const metaUrl = (Function('return import.meta.url')() as string) ?? '';
  if (metaUrl) return dirname(fileURLToPath(metaUrl));
  return process.cwd();
}

function resolveDefaultSpecDir(): string {
  const envDir = process.env[ENV_OVERRIDE];
  if (envDir && envDir.trim().length > 0) {
    return resolve(envDir);
  }

  const here = resolveModuleDir();

  const bundled = resolve(here, 'specs', 'v1');
  if (existsSync(bundled)) return bundled;

  const sourceTree = resolve(here, '..', '..', '..', '..', 'spec', 'management', 'v1');
  if (existsSync(sourceTree)) return sourceTree;

  throw new Error(
    `loadManagementSpecs: could not locate spec directory. Tried bundled ${bundled} and source ${sourceTree}. Set ${ENV_OVERRIDE} to override.`,
  );
}

function assertSpecShape(name: string, raw: unknown): asserts raw is ManagementSpec {
  if (!raw || typeof raw !== 'object') {
    throw new Error(`spec ${name}: not an object`);
  }
  const obj = raw as Record<string, unknown>;
  for (const required of ['name', 'description', 'inputSchema', 'outputSchema', 'since']) {
    if (!(required in obj)) {
      throw new Error(`spec ${name}: missing required field "${required}"`);
    }
  }
  if (typeof obj.name !== 'string' || typeof obj.description !== 'string') {
    throw new Error(`spec ${name}: name/description must be strings`);
  }
  if (typeof obj.since !== 'string') {
    throw new Error(`spec ${name}: since must be a string`);
  }
  if (typeof obj.inputSchema !== 'object' || typeof obj.outputSchema !== 'object') {
    throw new Error(`spec ${name}: inputSchema/outputSchema must be objects`);
  }
}

function readJson(path: string): unknown {
  return JSON.parse(readFileSync(path, 'utf-8'));
}

/**
 * Load every `<tool>.json` and matching `examples/<tool>.example.json`
 * fixture from the management spec directory.
 *
 * Resolution order for `specDir`:
 * 1. Explicit option, when passed
 * 2. `COPASS_MANAGEMENT_SPEC_DIR` env override
 * 3. Bundled `dist/specs/v1/`
 * 4. Source-tree `copass/spec/management/v1/` (dev mode)
 */
export function loadManagementSpecs(options: LoadOptions = {}): LoadedManagementCorpus {
  const specDir = options.specDir ? resolve(options.specDir) : resolveDefaultSpecDir();

  if (!existsSync(specDir) || !statSync(specDir).isDirectory()) {
    throw new Error(`loadManagementSpecs: ${specDir} is not a directory`);
  }

  const specs: Record<string, ManagementSpec> = {};
  for (const entry of readdirSync(specDir)) {
    if (!entry.endsWith('.json')) continue;
    const file = join(specDir, entry);
    if (!statSync(file).isFile()) continue;
    const raw = readJson(file);
    assertSpecShape(entry, raw);
    specs[raw.name] = raw;
  }

  const fixtures: Record<string, ManagementFixture> = {};
  const examplesDir = join(specDir, 'examples');
  if (existsSync(examplesDir) && statSync(examplesDir).isDirectory()) {
    for (const entry of readdirSync(examplesDir)) {
      if (!entry.endsWith('.example.json')) continue;
      const toolName = entry.slice(0, -'.example.json'.length);
      const file = join(examplesDir, entry);
      const raw = readJson(file) as Record<string, unknown>;
      if (!raw || typeof raw !== 'object' || !('input' in raw) || !('output' in raw)) {
        throw new Error(`fixture ${entry}: must have "input" and "output" keys`);
      }
      fixtures[toolName] = raw as unknown as ManagementFixture;
    }
  }

  return { specDir, specs, fixtures };
}

File: python/copass-management/src/copass_management/specs.py
"""Spec corpus loader.

Resolves the management spec directory in this order:

1. ``spec_dir`` keyword argument when passed.
2. ``COPASS_MANAGEMENT_SPEC_DIR`` env override.
3. Vendored copy bundled inside the package
   (``copass_management/_spec/v1/``).
4. Source tree
   (``copass/spec/management/v1/``) — only resolved during local
   dev when the package is installed editable from the harness checkout.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from importlib import resources
from pathlib import Path
from typing import Any, Dict, Iterable, Optional

MIN_SPEC_VERSION = "v1"
MAX_SPEC_VERSION = "v1"

_ENV_OVERRIDE = "COPASS_MANAGEMENT_SPEC_DIR"


@dataclass(frozen=True)
class ManagementSpec:
    name: str
    description: str
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any]
    since: str

    @classmethod
    def from_json(cls, data: Dict[str, Any]) -> "ManagementSpec":
        for required in ("name", "description", "inputSchema", "outputSchema", "since"):
            if required not in data:
                raise ValueError(f"spec {data.get('name', '?')!r} missing required field {required!r}")
        if not isinstance(data["name"], str) or not isinstance(data["description"], str):
            raise ValueError("spec name/description must be strings")
        if not isinstance(data["since"], str):
            raise ValueError("spec since must be a string")
        if not isinstance(data["inputSchema"], dict) or not isinstance(data["outputSchema"], dict):
            raise ValueError("inputSchema/outputSchema must be objects")
        return cls(
            name=data["name"],
            description=data["description"],
            input_schema=data["inputSchema"],
            output_schema=data["outputSchema"],
            since=data["since"],
        )


@dataclass(frozen=True)
class ManagementFixture:
    input: Dict[str, Any]
    output: Dict[str, Any]


@dataclass(frozen=True)
class LoadedManagementCorpus:
    spec_dir: Path
    specs: Dict[str, ManagementSpec] = field(default_factory=dict)
    fixtures: Dict[str, ManagementFixture] = field(default_factory=dict)


def _bundled_spec_dir() -> Optional[Path]:
    try:
        with resources.as_file(resources.files("copass_management") / "_spec" / "v1") as path:
            if path.exists():
                # `as_file` may yield a temp path; copy semantics are safe
                # because we only read.
                return Path(path)
    except (ModuleNotFoundError, FileNotFoundError):
        return None
    return None


def _source_tree_spec_dir() -> Optional[Path]:
    here = Path(__file__).resolve().parent
    candidate = here.parent.parent.parent.parent.parent / "spec" / "management" / "v1"
    if candidate.exists():
        return candidate
    return None


def _resolve_default_spec_dir() -> Path:
    env_override = os.environ.get(_ENV_OVERRIDE, "").strip()
    if env_override:
        return Path(env_override).resolve()

    bundled = _bundled_spec_dir()
    if bundled is not None:
        return bundled

    source = _source_tree_spec_dir()
    if source is not None:
        return source

    raise FileNotFoundError(
        "load_management_specs: could not locate spec directory. Set "
        f"{_ENV_OVERRIDE} to override."
    )


def _iter_specs(root: Path) -> Iterable[Path]:
    for entry in sorted(root.iterdir()):
        if entry.is_file() and entry.suffix == ".json":
            yield entry


def load_management_specs(*, spec_dir: Optional[Path] = None) -> LoadedManagementCorpus:
    """Load every ``<tool>.json`` and matching example fixture from the
    management spec directory."""
    root = Path(spec_dir).resolve() if spec_dir is not None else _resolve_default_spec_dir()
    if not root.is_dir():
        raise NotADirectoryError(f"load_management_specs: {root} is not a directory")

    specs: Dict[str, ManagementSpec] = {}
    for spec_file in _iter_specs(root):
        raw = json.loads(spec_file.read_text(encoding="utf-8"))
        spec = ManagementSpec.from_json(raw)
        specs[spec.name] = spec

    fixtures: Dict[str, ManagementFixture] = {}
    examples_dir = root / "examples"
    if examples_dir.is_dir():
        for fixture_file in sorted(examples_dir.iterdir()):
            if not (fixture_file.is_file() and fixture_file.name.endswith(".example.json")):
                continue
            tool_name = fixture_file.name[: -len(".example.json")]
            raw = json.loads(fixture_file.read_text(encoding="utf-8"))
            if not isinstance(raw, dict) or "input" not in raw or "output" not in raw:
                raise ValueError(f"fixture {fixture_file.name!r} must have 'input' and 'output' keys")
            fixtures[tool_name] = ManagementFixture(input=raw["input"], output=raw["output"])

    return LoadedManagementCorpus(spec_dir=root, specs=specs, fixtures=fixtures)


__all__ = [
    "MIN_SPEC_VERSION",
    "MAX_SPEC_VERSION",
    "ManagementSpec",
    "ManagementFixture",
    "LoadedManagementCorpus",
    "load_management_specs",
]
